lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -1356,11 +1355,119 @@ export class UserAgentApplication {
      */
     private getCachedToken(serverAuthenticationRequest: ServerRequestParameters, account: Account): AuthResponse {
         this.logger.verbose("GetCachedToken has been called");
-        let accessTokenCacheItem: AccessTokenCacheItem = null;
         const scopes = serverAuthenticationRequest.scopes;
 
+        /**
+         * Id Token should be returned in every acquireTokenSilent call. The only exception is a response_type = token
+         * request when a valid ID Token is not present in the cache.
+         */
+        const idToken = this.getCachedIdToken(serverAuthenticationRequest, account);
+        switch(serverAuthenticationRequest.responseType) {
+            case ResponseTypes.id_token:
+                this.logger.verbose("Returning ID Token AuthResponse cache result");
+                return this.buildIdTokenAuthResponse(idToken, serverAuthenticationRequest, scopes, account);
+            case ResponseTypes.id_token_token:
+                this.logger.verbose("Returning Access Token and ID Token AuthResponse cache result");
+                return this.buildIdTokenTokenAuthResponse(idToken, serverAuthenticationRequest, scopes, account);
+            default: // ResponseTypes.token
+                this.logger.verbose("Returning Access Token AuthResponse cache result");
+                return this.buildTokenAuthResponse(idToken, serverAuthenticationRequest, scopes, account);
+        }
+    }
+
+    private buildTokenAuthResponse(idToken: IdToken, serverAuthenticationRequest: ServerRequestParameters, scopes: Array<string>, account: Account): AuthResponse {
```

Comment:
Could we turn these into one generic static function on `ResponseUtils`?

Reply:
I tried at first, but the refactor wasn't trivial. There are calls to private methods like `getCachedToken()` and `getAccountState`. I'll call those methods before calling these utility functions and pass the results as arguments.

Reply:
Done!

Reply:
Oh sorry, misread the part about making it a single method. Let me pass the switch into a utility function and handle it in ResponseUtils.